minerl/Malmo/Minecraft/src/main/java/com/microsoft/Malmo/MissionHandlers/ObservationFromDamageImplementation.java

```diff
@@ -0,0 +1,142 @@
+        }
+    }
+
+    @SubscribeEvent
+    public void onDamage(LivingHurtEvent event) {
+        if (event.getEntityLiving().equals(Minecraft.getMinecraft().player)) {
+            if (this.damageSource != null) {
+                this.damageSourceReplaced = true;
+                // If multiple LivingHurtEvents pertain to the player, we keep the most recent one and record the fact
+                // that we are skipping a damage event in leiu of building an array based reporting system
+                System.out.println("Warning skipped damage event - multiple damage events in one tick!");
+            }
+            this.damageSource = event.getSource();
+            this.damageAmount = event.getAmount();
+            this.entity = event.getEntityLiving();
+        }
+    }
+
+    private void resetDamage() {
+        this.damageAmount = 0;
+        this.damageSource = null;
+        this.entity = null;
+        this.hasDied = false;
+    }
+
+    @Override
+    public void writeObservationsToJSON(JsonObject json, MissionInit missionInit) {
+        EntityPlayerSP player = Minecraft.getMinecraft().player;
+        JsonObject damage_json = new JsonObject();
+        json.addProperty("is_dead", player.isDead);
+        json.addProperty("living_death_event_fired", this.hasDied);
+
+        // Note to user that there could be inaccuracies here
+        if (this.damageSourceReplaced) {
+            damage_json.addProperty("stale", true);
+            this.damageSourceReplaced = false;
+        }
+
+        if (this.damageAmount != 0 && this.damageSource != null) {
+            System.out.println(this.damageAmount + " damage from " + this.damageSource.getDamageType() + " by entity " + this.damageSource.getEntity());
+        }
+
+        if (this.damageAmount != 0) {
+            damage_json.addProperty("damage_amount", this.damageAmount);
+        }
+
+        if (this.damageSource != null) {
+            damage_json.addProperty("damage_type", this.damageSource.getDamageType());
+            damage_json.addProperty("hunger_damage", this.damageSource.getHungerDamage());
+            damage_json.addProperty("is_damage_absolute", this.damageSource.isDamageAbsolute());
+            damage_json.addProperty("is_fire_damage", this.damageSource.isFireDamage());
+            damage_json.addProperty("is_magic_damage", this.damageSource.isMagicDamage());
+            damage_json.addProperty("is_difficulty_scaled", this.damageSource.isDifficultyScaled());
+            damage_json.addProperty("is_explosion", this.damageSource.isExplosion());
+            damage_json.addProperty("is_projectile", this.damageSource.isProjectile());
+            damage_json.addProperty("is_unblockable", this.damageSource.isUnblockable());
+            damage_json.addProperty("death_message", this.damageSource.getDeathMessage(player).getUnformattedText());
+
+            if (this.damageSource.getEntity() != null) {
+                Entity entity = this.damageSource.getEntity();
```

Comment:
FYI -- it looks like `this.entity` is unused and overshadowed by this local variable here.

I would recommend renaming one of these guys to something different, or maybe even deleting the unused `this.entity`.

Reply:
this.entity seems to only ever be the player - where as this.damageSource.getEntity() is the entity causing the damage - will update